c10/cuda/CUDAFunctions.cpp

```diff
@@ -93,6 +93,8 @@ DeviceIndex device_count() noexcept {
   // initialize number of devices only once
   static int count = []() {
     try {
+      auto result = device_count_impl();
+      TORCH_INTERNAL_ASSERT(result < (1 << (sizeof(DeviceIndex) * CHAR_BIT)), "Too many CUDA devices, DeviceIndex overflowed");
```

Comment:
technically it's a signed value, so you'd need to check 127? you can probably do `std::numeric_limits<DeviceIndex>::max` to avoid the bit shifts